### PR TITLE
DDG: Remove kind.server filter and validate the case of service calling itself

### LINKED ISSUE #554: DDG for search results is still wrong
The HotROD graph is missing DB nodes. The redis and mysql services do show in the search results, as well as in the regular DAG graph, but not in DDG:

image

### PR DESCRIPTION
**Pull request by rubenvp8510**
## Which problem is this PR solving?
- Fixes #554 

## Short description of the changes
- I removed the span.kind == server filter, instead validate when the service calls itself multiple times and remove all of those recursive calls and only preserve one of them.

Attached an image.

Screenshot from 2020-04-04 09-35-55

### CODE REVIEW

**Comment by rubenvp8510** [comment]
Still need to fix some tests.

File: packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx [hunk 1]
```diff
@@ -35,22 +38,33 @@ function transformTracesToPaths(
...
-            .filter(span => span.tags.find(({ key, value }) => key === 'span.kind' && value === 'server'))
-            .map(({ processID, operationName: operation }) => ({
-              service: data.processes[processID].serviceName,
-              operation,
-            }));
-
+            if (reducedSpans.length > 0) {
+              const headSpan = reducedSpans[reducedSpans.length - 1];
+              // Transition inside the same service ServiceA -> ServiceA
+              if (headSpan.processID === span.processID) {
```
Comment (everett980) [inline packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx:51]: @rubenvp8510 sorry I couldn't get to this sooner

If I'm reading this correctly, it captures every change in `processID`. Then, if there isn't a change in `processID` and the current span is of kind `server`:
1) a parent `client` span that was a different process than _its_ parent span gets replaced with the `server` span
2) a parent `server` span is kept and the current span is added.

I'm wondering if we need to do the replacement in case 1, or if it would be better to simply add any span of kind `server` or any span with a different `processID` than the previous reduced span.
i.e.:
```
if (headSpan.processID !== span.processID || isKindServer(span) {
  reducedSpans.push(span);
}
```
The only case this is different is if `process a` calls `process b` which emits a `client` span as its first span, which is the parent of a `server` span in the same process. IMO this case is strange, possibly caused by bad instrumentation, and I'd prefer to surface the scenario to potentially fix it rather than bury it.

@yurishkuro @vprithvi I'd like to hear your opinions on the two approaches to handling this case.

Comment (everett980) [reply packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx:51]: @rubenvp8510 Update: I have discussed the two approaches with Yuri and Prithvi and we agree to not replace the ingress client spans.

Here's an example, which, for brevity, I will represent a single span as `processID,operationName,[s]erver/[c]lient`
e.g.:
```
{
  processID: 0,
  operationName: 'a',
  span.kind: 'client',
}
```
would be `0,a,c`

original:
`0,a,c 1,a,s 1,b,c 1,c,s, 2,a,c, 2,b,s 2,c,c`
would be reduced to
`0,a,c 1,a,s 1,c,s 2,a,c 2,b,s` by my proposed method
versus
`0,a,c 1,a,s 1,c,s 2,b,s` in the PR

The differences is that the client ingress span (`2,a,c`) is not dropped.

Comment (rubenvp8510) [reply packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx:51]: Hi @everett980 I think it is OK not loses the ingress client span. it just wasn't clear to me what would be the correct way of derivative the paths on those cases, but now it makes more sense. I'll do the pertinent changes.

**Comment by rubenvp8510** [comment]
This also fixes https://github.com/jaegertracing/jaeger-ui/issues/546 I added a test for that case.

File: packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx (hunk 1 above)
Comment (rubenvp8510) [reply packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx]: I did the changes, this is ready for another review.

Thank you!

File: packages/jaeger-ui/src/model/ddg/transformTracesToPaths.test.js [hunk 2]
```diff
@@ -22,10 +22,18 @@ describe('transform traces to ddg paths', () => {
     })),
     attributes: [{ key: 'exemplar_trace_id', value: trace.data.traceID }],
   });
-  const makeSpan = (spanName, parent, kind) => ({
+
+  const addExemplarTraceIDs = (path, traces) => {
+    traces.forEach(trace => {
+      path.attributes.push({ key: 'exemplar_trace_id', value: trace.data.traceID });
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/model/ddg/transformTracesToPaths.test.js:28]: @everett980 do we support span IDs for DDG exemplars?

File: packages/jaeger-ui/src/model/ddg/transformTracesToPaths.test.js [hunk 3]
```diff
@@ -22,10 +22,18 @@ describe('transform traces to ddg paths', () => {
     })),
     attributes: [{ key: 'exemplar_trace_id', value: trace.data.traceID }],
   });
-  const makeSpan = (spanName, parent, kind) => ({
+
+  const addExemplarTraceIDs = (path, traces) => {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/model/ddg/transformTracesToPaths.test.js:26]: Why do we use `const`+lambda instead of `function`? I find function a lot more readable & explicit than a lambda.

File: packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx [hunk 4]
```diff
@@ -35,22 +38,27 @@ function transformTracesToPaths(
...
+              isKindServer(span)
+            ) {
+              reducedSpans.push(span);
+            }
+            return reducedSpans;
+          }, []);
+          const path: TDdgPayloadEntry[] = spans.map(({ processID, operationName: operation }) => ({
+            service: data.processes[processID].serviceName,
+            operation,
+          }));
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx:61]: I find this algorithm quite hard to follow, not to mention that it seems to do a lot of redundant work by constantly recalculating the paths from each leaf upwards over and over. Have we tried implementing it as a classic BFS on a tree?

Comment (rubenvp8510) [reply packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx:61]: We can do that, but one think I would like to avoid is using recursion, it will lead to bad performance (see https://github.com/jaegertracing/jaeger-ui/pull/326).

Can we do that in a separate PR?

Comment (yurishkuro) [reply packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx:61]: ok

File: packages/jaeger-ui/src/model/ddg/transformTracesToPaths.test.js (hunk 2 above)
Comment (everett980) [reply packages/jaeger-ui/src/model/ddg/transformTracesToPaths.test.js:28]: @yurishkuro not yet. I asked @vprithvi about it but it wasn't a priority at the time.
Once my inflight work for the trace quality v2 page is landed, it'll be easier to add to the DDG as that page has many deep links and deep link utils.
However url length restrictions are already hit without deep linking, and deep linking would double the length so even then it won't be a super trivial change.
Let's create a ticket for it and put a pin it for now.

File: packages/jaeger-ui/src/model/ddg/transformTracesToPaths.test.js (hunk 3 above)
Comment (rubenvp8510) [reply packages/jaeger-ui/src/model/ddg/transformTracesToPaths.test.js:26]: I just followed the other parts of this code, I think there is no special reason for using lambda.

File: packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx (hunk 4 above)
Comment (rubenvp8510) [reply packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx:61]: I created another version of this algorithm using a depth breadth search tree algorithm, I'll update this PR soon.

Comment (rubenvp8510) [reply packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx]: Updated the algorithm, I haven't reused the tree , I can do it in another PR if it's required.

File: packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx [hunk 5]
```diff
@@ -30,53 +32,55 @@ function transformTracesToPaths(
...
-                service === focalService && (!focalOperation || operation === focalOperation)
-            )
+      data.spans.forEach(span => spanMap.set(span.spanID, span));
+      const tree = getTraceSpanIdsAsTree(data);
+      tree.paths((pathIds: string[]) => {
+        const paths = pathIds.reduce((reducedSpans: Span[], id: string): Span[] => {
+          if (id === '__root__') {
+            return reducedSpans;
+          }
+          const span = spanMap.get(id);
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx:42]: Q: why does `tree.paths()` generate paths as `ids[]` instead of `spans[]`? The tree descent does operate on spans, and the paths generator does not know what the paths will be used for, so returning IDs instead of spans seems unnecessary, only requiring you to re-map IDs back to spans.

Comment (rubenvp8510) [reply packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx:42]: This is a good point, and actually the reason I didn't do in that way is that I'm using the same tree used in transform-trace-data, which in each node only have the `spanID`.

We can of course change that to store the whole span on the node, but that will require other changes.

Comment (yurishkuro) [reply packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx:42]: fair enough

File: packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx [hunk 6]
```diff
@@ -30,53 +32,55 @@ function transformTracesToPaths(
...
-          if (
-            path.some(
-              ({ service, operation }) =>
-                service === focalService && (!focalOperation || operation === focalOperation)
-            )
+      data.spans.forEach(span => spanMap.set(span.spanID, span));
+      const tree = getTraceSpanIdsAsTree(data);
+      tree.paths((pathIds: string[]) => {
+        const paths = pathIds.reduce((reducedSpans: Span[], id: string): Span[] => {
+          if (id === '__root__') {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx:39]: please use `TREE_ROOT_ID` constant from `packages/jaeger-ui/src/selectors/trace.js`

File: packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx [hunk 7]
```diff
@@ -30,53 +32,55 @@ function transformTracesToPaths(
...
+            return reducedSpans;
+          }
+          const span = spanMap.get(id);
+          if (!span) throw new Error(`Ancestor spanID ${id} not found in trace ${traceID}`);
+          if (reducedSpans.length === 0) {
+            reducedSpans.push(span);
+          } else if (
+            reducedSpans[reducedSpans.length - 1].processID !== span.processID ||
+            isKindServer(span)
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx:48]: the whole algorithm expressed in these two lines, awesome!

 cc @vprithvi

File: packages/jaeger-ui/src/model/ddg/transformTracesToPaths.test.js [hunk 8]
```diff
@@ -52,7 +60,7 @@ describe('transform traces to ddg paths', () => {
         (result, span) => ({
           ...result,
           [span.processID]: {
-            serviceName: `${span.spanID.split(' ')[0]} service`,
+            serviceName: span.processID,
```
Comment (everett980) [inline packages/jaeger-ui/src/model/ddg/transformTracesToPaths.test.js:63]: This doesn't fully test that the `serviceName`s are populated. change this to
```
serviceName: ${span.processID}-name,
```

Comment (rubenvp8510) [reply packages/jaeger-ui/src/model/ddg/transformTracesToPaths.test.js]: Done. 

Thanks for the review!
